joomla/joomla-cms
# Pre-Upgrade Compatibility Check on Plugins That Could Kill Joomla Upgrade

Pull Request for Issue #31042   (edited to add testing packages)

### Summary of Changes
New method in JoomlaupdateModelDefault -> getNonCorePlugins which fetches a list of plugins filtered by optional list of plugin folders.  The returning array excludes core Joomla plugins.
PreUpdate Check now fetches a list of 'system','user','authentication','actionlog' and 'twofactorauth' plugins that could, conceviably, break a Joomla upgrade if they contain code that isn't compatible with the targetted Joomla version
The PreUpdateChecker javascript now:

1. Disables the live update submit button until the update checker has finished or the user checks a checkbox to confirm they understand the risks of upgrading without performing appropriate checks/system changes
2. Checks the list of nonCorePlugins (mentioned above) to see if they are 'compatible' with the Joomla upgrade (or are part of a 'compatible' package).  Note that 'compatible' is based on results from the extension update server.
3. After the update check has run if there are no potential incompatible plugins the confirmation checkbox is removed
4. Additionally we now report results from failed requests to update servers (timeouts, 403, 500 errors etc.) separately

The messages displayed are changed as the update checker is run.

### Testing Instructions

In a J3.10 dev site install some extensions that are NOT compatible with Joomla 4 - at least one out of date version of an extension and one where there is no published J4 compatible version. Ideally you should have one with an invalid update server entry too.

I am attaching some test packages you can use to check - make sure to enable the system plugins after installing these test extensions. The components and plugins do nothing.
[testuptodatepackage.zip](https://github.com/joomla/joomla-cms/files/5421676/testuptodatepackage.zip)
[testuxpackage.zip](https://github.com/joomla/joomla-cms/files/5421677/testuxpackage.zip)
[testwuj3ej4package.zip](https://github.com/joomla/joomla-cms/files/5421678/testwuj3ej4package.zip)
[testwuj3n4package.zip](https://github.com/joomla/joomla-cms/files/5421679/testwuj3n4package.zip)
[testwuvcpackage.zip](https://github.com/joomla/joomla-cms/files/5421680/testwuvcpackage.zip)
[testxupackage.zip](https://github.com/joomla/joomla-cms/files/5421681/testxupackage.zip)

Then open Joomla Updater component and observe the pre-update checker tab and the live update tab - as the update checker is running and after it has completed

### Actual result BEFORE applying this Pull Request

You would be able to upgrade Joomla with an incompatible plugin installed that could 'brick' (i.e. kill) your Joomla installation which would involve a lot of work to resolve.

### Expected result AFTER applying this Pull Request

The modified update checker will stop you from performing the upgrade unless you have no incompatble plugins OR you click a checkbox to say you know what you are doing.

### Documentation Changes Required

Once the  PR is finalised the update checker and update process pages will need to be updated.
https://docs.joomla.org/Pre-Update_Check
